### PR TITLE
Align viewbox center with vacuum map center

### PR DESCRIPTION
**Pull request by philek**
This is an alternative approach for supporting the Vacuum Map card calibration from https://github.com/DeebotUniverse/Deebot-4-Home-Assistant/issues/321. This will make the calibration much simpler. Object coordinates inside the SVG are vacuum coordinates divided by 50 with y axis flipped. That information, combined with the viewbox, will be enough to translate the vacuum coordinates to a position on even a scaled image on the frontend.

This has the benefit over my previous PR of not changing the interface. The calibration points can be static and added outside of this library.

## Summary by CodeRabbit

- **New Features**
  - Improved map visualization with enhanced precision using the new `ViewBoxFloat` class.

- **Bug Fixes**
  - Enhanced accuracy in map subset, trace, and position calculations.

- **Tests**
  - Added and updated tests to cover new methods and ensure reliability of map functionality.

- **Refactor**
  - Streamlined code for better performance and maintainability in map-related operations.

### CODE REVIEW

**Comment by edenhaus** [comment]
Thanks for the PR. I like this more than your previous one :)
Can you please link to the PR, which implements the calibration points in the custom card?

**Comment by philek** [comment]
Rebased on latest dev.

**Comment by edenhaus** [comment]
As this PR is only needed for PiotrMachowski/lovelace-xiaomi-vacuum-map-card. What's the status there? Can the card used with these changes only, or does the card itself need changes?

**Comment by edenhaus** [comment]
@coderabbitai review

File: tests/test_map.py [hunk 1]
```diff
@@ -2,22 +2,10 @@
...
-    HorizontalLineToRel,
-    LineToRel,
-    MoveTo,
-    MoveToRel,
-    PathData,
-    Polygon,
-    SmoothCubicBezierRel,
-    VerticalLineToRel,
-)
+import svg
```
Comment (edenhaus) [inline tests/test_map.py:8]: Can you please revert this change.

Comment (philek) [reply tests/test_map.py]: Of course. Done.

**Comment by philek** [comment]
> As this PR is only needed for PiotrMachowski/lovelace-xiaomi-vacuum-map-card. What's the status there? Can the card used with these changes only, or does the card itself need changes?

In its current state it's very hard to calibrate the card correctly. With these changes it can be done quite easily by taking the min_x, min_y from the viewbox, even if done manually. I'm focusing on this part first.

File: tests/test_map.py [hunk 2]
```diff
@@ -2,7 +2,7 @@
 
 import asyncio
 from typing import TYPE_CHECKING
-from unittest.mock import ANY, AsyncMock, Mock, call
+from unittest.mock import ANY, AsyncMock, Mock, call, patch
```
[the remark this answers is not in the document]
Comment (edenhaus) [reply tests/test_map.py:5]: @coderabbitai Mocks are allowed to use

File: deebot_client/map.py [hunk 3]
```diff
@@ -237,33 +211,27 @@ class BackgroundImage:
...
+    point = _calc_point(x, y)
+    return Point(
+        min(
+            max(point.x, float(viewbox.min_x)),
+            float(viewbox.min_x) + float(viewbox.width),
+        ),
+        min(
+            max(point.y, float(viewbox.min_y)),
+            float(viewbox.min_y) + float(viewbox.height),
+        ),
```
[the remark this answers is not in the document]
Comment (edenhaus) [reply deebot_client/map.py:234]: Show us how you would simplify it

File: tests/test_map.py [hunk 4]
```diff
@@ -251,20 +269,58 @@ def test_points_to_svg_path(
...
-            svg_max=1000,
-        ),
-    )
-    assert _get_svg_subset(subset, manipulation) == expected
+    assert _get_svg_subset(subset) == expected
+
+
+_test_get_svg_positions_data = [
+    (
+        [Position(PositionType.CHARGER, 5000, -55000)],
```
Comment (edenhaus) [inline tests/test_map.py:283]:
```suggestion
        [Position(PositionType.CHARGER, 5000, -55000, 0)],
```

File: tests/test_map.py [hunk 5]
```diff
@@ -251,20 +269,58 @@ def test_points_to_svg_path(
...
+        ViewBoxSpec(-500, -500, 1000, 1000),
+        [Use(href="#c", x=100, y=500)],
+    ),
+    (
+        [Position(PositionType.DEEBOT, 15000, 15000)],
```
Comment (edenhaus) [inline tests/test_map.py:288]:
```suggestion
        [Position(PositionType.DEEBOT, 15000, 15000, 0)],
```

File: tests/test_map.py [hunk 6]
```diff
@@ -251,20 +269,58 @@ def test_points_to_svg_path(
...
+        ViewBoxSpec(-500, -500, 1000, 1000),
+        [Use(href="#d", x=300, y=-300)],
+    ),
+    (
+        [
+            Position(PositionType.CHARGER, 25000, 55000),
+            Position(PositionType.DEEBOT, -5000, -50000),
```
Comment (edenhaus) [inline tests/test_map.py:295]:
```suggestion
            Position(PositionType.CHARGER, 25000, 55000, 0),
            Position(PositionType.DEEBOT, -5000, -50000, 0),
```

File: tests/test_map.py [hunk 7]
```diff
@@ -251,20 +269,58 @@ def test_points_to_svg_path(
...
+        ],
+        ViewBoxSpec(-500, -500, 1000, 1000),
+        [Use(href="#d", x=-100, y=500), Use(href="#c", x=500, y=-500)],
+    ),
+    (
+        [
+            Position(PositionType.DEEBOT, -10000, 10000),
+            Position(PositionType.CHARGER, 50000, 5000),
```
Comment (edenhaus) [inline tests/test_map.py:303]:
```suggestion
            Position(PositionType.DEEBOT, -10000, 10000, 0),
            Position(PositionType.CHARGER, 50000, 5000, 0),
```

File: deebot_client/map.py [hunk 8]
```diff
@@ -38,7 +39,6 @@
 from .util import (
     OnChangedDict,
     OnChangedList,
-    decompress_7z_base64_data,
```
Comment (edenhaus) [inline deebot_client/map.py:41]: Please revert this change

File: deebot_client/map.py [hunk 9]
```diff
@@ -17,6 +17,7 @@
 from PIL import Image, ImageColor, ImageOps, ImagePalette
 import svg
 
+from deebot_client import util
```
Comment (edenhaus) [inline deebot_client/map.py:20]: Please revert this change

File: deebot_client/map.py [hunk 10]
```diff
@@ -295,22 +263,24 @@ def _points_to_svg_path(
...
     for position in sorted(positions, key=lambda x: _POSITIONS_SVG[x.type].order):
-        pos = _calc_point(position.x, position.y, map_manipulation)
+        pos = _calc_point_in_viewbox(position.x, position.y, viewbox)
         svg_positions.append(
-            svg.Use(href=f"#{_POSITIONS_SVG[position.type].svg_id}", x=pos.x, y=pos.y)
+            svg.Use(
+                href=f"#{_POSITIONS_SVG[position.type].svg_id}",
+                x=pos.x,
+                y=pos.y,
+            )
```
Comment (edenhaus) [inline deebot_client/map.py:276]:
```suggestion
            svg.Use(href=f"#{_POSITIONS_SVG[position.type].svg_id}", x=pos.x, y=pos.y)
```

File: deebot_client/map.py [hunk 11]
```diff
@@ -416,7 +385,7 @@ async def on_map_subset(event: MapSubsetEvent) -> None:
 
     def _update_trace_points(self, data: str) -> None:
         _LOGGER.debug("[_update_trace_points] Begin")
-        trace_points = decompress_7z_base64_data(data)
+        trace_points = util.decompress_7z_base64_data(data)
```
Comment (edenhaus) [inline deebot_client/map.py:388]:
```suggestion
        trace_points = decompress_7z_base64_data(data)
```

File: deebot_client/map.py [hunk 12]
```diff
@@ -667,7 +612,7 @@ def image(self) -> Image.Image:
 
     def update_points(self, base64_data: str) -> None:
         """Add map piece points."""
-        decoded = decompress_7z_base64_data(base64_data)
+        decoded = util.decompress_7z_base64_data(base64_data)
```
Comment (edenhaus) [inline deebot_client/map.py:615]:
```suggestion
        decoded = decompress_7z_base64_data(base64_data)
```

**Review by edenhaus — CHANGES_REQUESTED** [review]

**Comment by edenhaus** [comment]
I tested your change successfully, and the map looks correct :) Waiting for the last tweaks so I can merge it

**Review by edenhaus — APPROVED** [review]
Thanks @philek :+1:

File: tests/test_map.py [hunk 13]
```diff
@@ -157,7 +158,7 @@ async def on_change() -> None:
 
 
 @patch(
-    "deebot_client.util.decompress_7z_base64_data",
+    "deebot_client.map.decompress_7z_base64_data",
```
Comment (philek) [inline tests/test_map.py:161]: I didn't know you could do that. That does make it simpler :)

**Comment by philek** [comment]
Thank you for finishing it off :)

**Pull request merged**